Support for OIDC Logout

jzheaux opened the pull request:
This commit introduces two new classes for coordinating logout with
an OP. The first is for the user-agent (redirect) use case and the
second is for the server-side use case.

For now they are separate due to settings that don't make sense in
one use case vs the other, however, there is some duplicative logic
that we could either clean up with an abstract class or some
additional configuration logic.

This commit doesn't propose any changes to the OAuth2LoginConfigurer
as I'd first like to get feedback on the classes themselves.

Issue: gh-5350

jzheaux commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/logout/OidcEndpointLogoutHandler.java:
```diff
@@ -0,0 +1,133 @@
+
+	private boolean includeIdTokenHint;
+
+	private String oidcLogoutEndpointUrl;
```
This needs to pull configuration from client registrations as opposed to being hardcoded. I still need to take a look at ClientRegistrationRepository.

jzheaux replied:
Also, note that there is actually no place yet via client registration to specify this endpoint.

jzheaux commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/logout/OidcEndpointLogoutHandler.java:
```diff
@@ -0,0 +1,133 @@
+public class OidcEndpointLogoutHandler implements LogoutHandler {
+	private final Logger log = LoggerFactory.getLogger(this.getClass());
+
+	private final RestTemplate restTemplate;
```
RestOperations may be preferable since it is an interface.

jzheaux commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/logout/OidcRedirectingLogoutSuccessHandler.java:
```diff
@@ -0,0 +1,111 @@
+ * @since 5.1
+ * @see <a href="http://openid.net/specs/openid-connect-session-1_0.html#RPLogout">RP-Initiated Logout</a>
+ */
+public class OidcRedirectingLogoutSuccessHandler implements LogoutSuccessHandler {
```
We do have a SimpleUrlLogoutSuccessHandler, and it may be worth it to extend from that.

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/logout/OidcLogoutSuccessHandler.java:
```diff
@@ -0,0 +1,89 @@
+			return builder.encode(StandardCharsets.UTF_8).build().toUriString();
+		}
+
+		return getDefaultTargetUrl();
```
I think this should be `super.determineTargetUrl(request, response)` ? Also, I would prefer this to happen in the pre-checks to avoid the nested `if` - I find it more readable that way.

jzheaux commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/logout/OidcLogoutSuccessHandler.java:
```diff
@@ -0,0 +1,89 @@
+ * @see <a href="http://openid.net/specs/openid-connect-session-1_0.html#RPLogout">RP-Initiated Logout</a>
+ * @see {@link org.springframework.security.web.authentication.logout.LogoutSuccessHandler}
+ */
+public class OidcLogoutSuccessHandler extends SimpleUrlLogoutSuccessHandler {
```
Agreed, I like your first suggestion.

jgrandja replied:
I'm wondering if `OidcProviderLogoutSuccessHandler` makes more sense...since we're logging out at the provider?

jzheaux replied:
I like the `ClientInitiated` part because it ties us back to the spec.

Also, when I read `OidcProvider`, I hear something that belongs on the provider side, e.g. something that is "provider-initiated".

Aren't we always logging out at the provider (eventually)? Seems like we differentiate by saying where it starts.

jzheaux commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/logout/OidcLogoutSuccessHandler.java:
```diff
@@ -0,0 +1,89 @@
+	protected String determineTargetUrl(HttpServletRequest request,
+			HttpServletResponse response, Authentication authentication) {
+
+		if (authentication instanceof OAuth2AuthenticationToken) {
```
This happens below when deciding whether to include the `id_token_hint`. Are you saying, then, that this handler should basically be a pass-through if the user isn't an `OidcUser`?

jgrandja replied:
I understand this check is happening later on in the code but it should happen first. For example, if this handler is configured and a user logs in using a standard oauth provider (e.g. GitHub) than we should default to current behaviour.

> Are you saying, then, that this handler should basically be a pass-through

If the current authentication is not OIDC, than we should default to the behaviour that's happening today after a user logs out.

jzheaux commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/logout/OidcLogoutSuccessHandler.java:
```diff
@@ -0,0 +1,89 @@
+			ClientRegistration registration =
+					this.repository.findByRegistrationId(token.getAuthorizedClientRegistrationId());
+
+			String endpointUri = (String)
```
shoot, silly me

jzheaux commented:
@jgrandja, thanks for all the great feedback. I've applied it and also added its reactive counterpart.

I added the reactive at this point since it seems like we are close enough on the servlet code.

Also, note that I stuck with your first suggestion of `ClientInitiated` - I think it lines up nicely with the spec, see how nicely it reads alongside the javadoc for that class.

Lastly, @rwinch when you get a moment I'd love to get your take on the reactive code. Of particular note is that I duplicate some functionality already found in `RedirectServerLogoutSuccessHandler` because I couldn't get a delegate to work.

rwinch commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/logout/OidcClientInitiatedServerLogoutSuccessHandler.java:
```diff
@@ -0,0 +1,123 @@
+				.map(OAuth2AuthenticationToken.class::cast)
+				.flatMap(this::endSessionEndpoint)
+				.map(endSessionEndpoint -> endpointUri(endSessionEndpoint, authentication))
+				.defaultIfEmpty(this.logoutSuccessUrl)
```
I think we could create a member (please consider renaming but illustrates point):

```java
private ServerLogoutSuccessHandler delegate = new RedirectServerLogoutSuccessHandler();

// add setter
```

You can then remove `this.logoutSuccessUrl` and change `.defaultIfEmpty(this.logoutSuccessUrl)` to

```java
.switchIfEmpty(delegate.onLogoutSuccess(exchange, authentication).then(Mono.empty()))
```

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/logout/OidcClientInitiatedLogoutSuccessHandler.java:
```diff
@@ -0,0 +1,98 @@
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.client.oidc.logout;
```
@rwinch Please see my latest comment on package name as I believe the one I suggested is correct?

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/logout/OidcClientInitiatedServerLogoutSuccessHandler.java:
```diff
@@ -0,0 +1,123 @@
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.client.oidc.logout;
```
@rwinch Please see my latest comment on package name as I believe the one I suggested is correct?
